src/python-bindings/htcondor2/credd.cpp

```diff
@@ -1 +1,175 @@
+    // However, the dprint() logs will still differ from version 1.
+    //
+    // int result = do_store_cred(cooked_user.c_str(), credential, mode, & d);
+
+    ClassAd * serviceAd = NULL;
+    if( service != NULL ) {
+        serviceAd = new ClassAd();
+
+        serviceAd->Assign("service", service);
+        if( handle != NULL ) {
+            serviceAd->Assign("handle", handle);
+        }
+    } else if( handle != NULL ) {
+        // This was HTCondorValueError in version 1.
+        PyErr_SetString( PyExc_ValueError, "invalid service argument" );
+        return NULL;
+    }
+
+    Daemon * d = NULL;
+    if( addr != NULL ) {
+        d = new Daemon(DT_CREDD, addr);
+    }
+
+    ClassAd returnAd;
+    int result = do_store_cred( cooked_user.c_str(), mode,
+        (const unsigned char *)credential, credential_len,
+        returnAd, serviceAd, d);
+    if( d != NULL ) { delete d; }
+    if( serviceAd != NULL ) { delete serviceAd; }
+
+    const int MODE_MASK = 3;
+    bool query_mode = (mode & MODE_MASK) == GENERIC_QUERY;
+    bool delete_mode = (mode & MODE_MASK) == GENERIC_DELETE;
+    if( result == FAILURE_NOT_FOUND && (query_mode || delete_mode) ) {
+        return PyLong_FromLong(result);
+    }
+
+    const char * errString = NULL;
+    if( _store_cred_failed(result, mode, & errString) ) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, errString );
+        return NULL;
+    }
+
+    const int TYPE_MASK = 0x2c;
+    if( (mode & MODE_MASK) == GENERIC_QUERY &&
+      (mode & TYPE_MASK) == STORE_CRED_USER_OAUTH ) {
+        std::string str;
+        sPrintAd(str, returnAd);
+        return PyUnicode_FromString(str.c_str());
+    }
+
+    return PyLong_FromLong(result);
+}
+
+
+static PyObject *
+_credd_do_check_oauth_creds(PyObject *, PyObject * args) {
+    // _credd_do_check_oauth_creds(addr, user, serviceAds)
+
+    const char * addr = NULL;
+    const char * user = NULL;
+    long mode = 0;
+    PyObject * pyServiceAds = NULL;
+
+    if(! PyArg_ParseTuple( args, "zzlO", & addr, & user, & mode, & pyServiceAds )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    std::string cooked_user;
+    if(! cook_user(user, mode, cooked_user)) {
+        // This was HTCondorValueError in version 1.
+        PyErr_SetString( PyExc_ValueError, "invalid user argument" );
+        return NULL;
+    }
+
+    Daemon * d = NULL;
+    if( addr != NULL ) {
+        d = new Daemon(DT_CREDD, addr);
+    }
+
+    std::vector<const ClassAd *> requests;
+    Py_ssize_t size = PyList_Size( pyServiceAds );
+    for( int i = 0; i < size; ++i ) {
+        // We verified the types of these on the Python side.
+        PyObject * py_classad = PyList_GetItem( pyServiceAds, i );
+        auto * handle = get_handle_from( py_classad );
+        ClassAd * classad = (ClassAd *)handle->t;
+
+        ClassAd * copy = new ClassAd();
+        copy->CopyFrom(* classad);
```

Comment:
Is making a copy of each ClassAd required. Version 1 doesn't do that.

Reply:
I don't actually know.  What I do know is that this guarantees that the C++ code won't alter or attempt to free the C++ `ClassAd` objects owned by the Python ClassAd objects.  I understand the request ads to be few and small, so even if this is unnecessary, it shouldn't be a performance problem.

Reply:
do_check_oauth_creds() doesn't mutate the ads. It's making its own copy of them to mutate. I'm ok leave this extra copy in place for now.